File: Python/LeetCode/Num242.py
"""
Date:
  23.05.12
Number:
  LeetCode 242
Title:
  Valid Anagram
Level:
  Easy
Name:
  thelight0804
"""

#### use hash table
class Solution(object):
  def isAnagram(self, s, t):
    """
    :type s: str
    :type t: str
    :rtype: bool
    """
    table = {}

    # add character count of c
    for c in s:
      if table.get(c) is None:
        table[c] = 1
      else:
        table[c] += 1
    
    # sub character count of t
    for c in t:
      #not exist character
      if table.get(c) is None:
        return False
      else:
        table[c] -= 1
        #not match character count
        if table[c] < 0: return False
    
    #not match character count
    for i in table.values():
      if i != 0: return False
    
    #match anagram
    return True

#### use sort
class Solution(object):
  def isAnagram(self, s, t):
    """
    :type s: str
    :type t: str
    :rtype: bool
    """
    if len(s) != len(t): return False
    elif sorted(s) != sorted(t): return False
    else: return True

File: JavaScript/LeetCode/Num242.js
/**
 * Date: 23.10.17
 * Number: LeetCode 242
 * Title: Valid Anagram
 * Level: Easy
 * Author: thelight0804
 */


// Use sort
/**
 * @param {string} s
 * @param {string} t
 * @return {boolean}
 */
var isAnagram = function(s, t) {
    // chech two strings length
    if (s.length !== t.length) return false;

    // sort string in ascending order 
    s = s.split('').sort().join('');
    t = t.split('').sort().join('');
    
    // check same string
    return s===t;
};

// Use hash table
var isAnagram = function(s, t) {
    // chech two strings length
    if (s.length !== t.length) return false;

    // Create Hash map(hash table)
    var map = new Map();

    // Traverse s character
    for (var char of s){
        // Add count of character
        var count = map.get(char);
        if (map.get(char)){
            map.set(char, ++count);
        }
        else {
            // key: character, value: count
            map.set(char, 1);
        }
    }
    
    // Traverse t character
    for (var char of t){
        var count = map.get(char);
        // count is undefined or 0, is not Anagram
        if (!count) return false;
        // decrease count
        else map.set(char, --count);
    }

    // s and t is Anagram
    return true;
};
